cranelift/codegen/src/machinst/blockorder.rs

```diff
@@ -110,390 +92,196 @@ pub struct BlockLoweringOrder {
-            let block_succ_start = block_succs.len();
-            let mut succ_idx = 0;
-            visit_block_succs(f, block, |inst, succ, from_table| {
+            let start = block_succs.len();
+            visit_block_succs(f, block, |_, succ, from_table| {
                 block_out_count[block] += 1;
                 block_in_count[succ] += 1;
-                block_succs.push((inst, succ_idx, succ));
-                succ_idx += 1;
+                block_succs.push(LoweredBlock::Orig { block: succ });
 
                 if from_table {
                     indirect_branch_target_clif_blocks.insert(succ);
                 }
             });
-            let block_succ_end = block_succs.len();
-            block_succ_range[block] = (block_succ_start, block_succ_end);
 
+            // Ensure that blocks terminated by br_table instructions with an empty jump table are
+            // still treated like conditional blocks from the point of view of critical edge
+            // splitting.
             if let Some(inst) = f.layout.last_inst(block) {
-                if f.dfg.insts[inst].opcode() == Opcode::Return {
-                    // Implicit output edge for any return.
-                    block_out_count[block] += 1;
-                }
-            }
-        }
-        // Implicit input edge for entry block.
-        block_in_count[entry] += 1;
-
-        // All blocks ending in conditional branches or br_tables must
-        // have edge-moves inserted at the top of successor blocks,
-        // not at the end of themselves. This is because the moves
-        // would have to be inserted prior to the branch's register
-        // use; but RA2's model is that the moves happen *on* the
-        // edge, after every def/use in the block. RA2 will check for
-        // "branch register use safety" and panic if such a problem
-        // occurs. To avoid this, we force the below algorithm to
-        // never merge the edge block onto the end of a block that
-        // ends in a conditional branch. We do this by "faking" more
-        // than one successor, even if there is only one.
-        //
-        // (One might ask, isn't that always the case already? It
-        // could not be, in cases of br_table with no table and just a
-        // default label, for example.)
-        for block in f.layout.blocks() {
-            if let Some(inst) = f.layout.last_inst(block) {
-                // If the block has a branch with any "fixed args"
-                // (not blockparam args) ...
-                if f.dfg.insts[inst].opcode().is_branch() && f.dfg.inst_fixed_args(inst).len() > 0 {
-                    // ... then force a minimum successor count of
-                    // two, so the below algorithm cannot put
-                    // edge-moves on the end of the block.
-                    block_out_count[block] = std::cmp::max(2, block_out_count[block]);
+                if Opcode::BrTable == f.dfg.insts[inst].opcode() {
```

Comment:
This isn't quite the same as the existing logic, though today it evaluates the same: the existing logic is that if there are any fixed args (ie, non-blockparam args), we need to ensure no edge-moves to succs end up in this block. Branch tables with empty tables can result in this, but I'd prefer to encode the actual restriction for robustness (in other words, I want correctness written explicitly rather than an emergent property). Can we take the `f.dfg.insts[inst].opcode().is_branch() && f.dfg.inst_fixed_args(inst).len() > 0` condition and use it here?

Reply:
The third option, just to record it for completeness here, is to make RA2 handle this edge (pun intended) case properly, by considering a branch's uses to come after edge-moves if they're inserted before the branch. I'm not advocating for that by any means, but I just wanted to note that end of the spectrum exists and is, in some manner of thinking, the cleanest :-)

Thanks for writing out the reasoning here, in more depth than I had considered -- given that we have the backstop of the checks in RA2 itself, running during fuzzing, I'm fine with this as-written for now. So, LGTM with just the above `succ_idx` change.